bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/internal/ui/navigator/SymbolsManager.java

```diff
@@ -228,8 +230,8 @@ private void refreshTreeContentFromLS(CompileUnit compileUnit) {
 		if (compileUnit == null || !compileUnit.isDirty) {
 			return;
 		}
+		lock.lock();
```

Comment:
So, we want other threads wait here and then execute the rest of the method. Is this true @ghentschke ?

Or do we want other threads to exit quickly if refresh is already in progress? In this case we should consider using `tryLock` instead.

Reply:
> So, we want other threads wait here and then execute the rest of the method. Is this true @ghentschke ?

yes, that's what I had in my mind. The (not very common) use case is: a user expands several TU right after another:
![image](https://github.com/eclipse-cdt/cdt-lsp/assets/123444711/e647a438-e265-443e-8681-2fffc5ef5421)
The lock shall ensure that they are executed successively without any mutual reactions.